A new air to water heat pump object

yujiex opened the pull request:
Pull request overview
---------------------


<!-- NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE -->

### Description of the purpose of this PR

Heat pumps are an essential technology for decarbonizing buildings. EnergyPlus is capable of modeling several types of heat pumps, e.g., for space cooling or heating: air to air, air to water, water to water; and heat pump water heaters. Although there is an existing feature to model air-to-water heat pumps in EnergyPlus, its capability is limited and its implementation is not straightforward compared with other types of HP. Some of the limitations include: 
- Cannot specify parameters like single, multi-speed, or variable speed for capacity control (for example, the Daikin model EWYT-CZ series is inverter controlled, and the EWYQ series is a fixed-speed two-stage model.)
- Missing details of crankcase heater control and more flexible defrost control
- The naming of the object is not intuitive - it does not refer to water to water or air to water, making searching for the object difficult
- Missing the input of an availability schedule (which disables the HP for seasonal consideration)
- Designed as separate heating and cooling objects while in reality, it should be one piece of equipment

This feature will create a new air-to-water heat pump (AWHP) model that includes heating and cooling components in one single object, capable of multi-speed controls, and includes more equipment details such as fan characteristics, PLR ranges, availability schedules, etc.

#### About the extended internal data structure arrays

The following entried are added to the internal data structure to hold info about this new AWHP.

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-c5b4a282-7fff-0cfa-9d28-57b91830c636"><div dir="ltr" style="margin-left:72pt;" align="left">
Entries | Added to
-- | --
HeatPumpAirToWaterCooling,<br>HeatPumpAirToWaterHeating,<br>HeatPumpAirToWater, | “ConnectionObjectType” in DataLoopNodes.hh<br>“PlantEquipmentType” in Enums.hh
"HeatPump:AirToWater:Cooling",<br>"HeatPump:AirToWater:Heating",<br>"HeatPump:AirToWater", | PlantEquipTypeNames in DataPlant.hh<br>“ConnectionObjectTypeNames” in BranchNodeConnections.cc
"HEATPUMP:AIRTOWATER:COOLING",<br>"HEATPUMP:AIRTOWATER:HEATING",<br>"HEATPUMP:AIRTOWATER", | “PlantEquipTypeNamesUC” in DataPlant.hh<br>“ConnectionObjectTypeNamesUC” in BranchNodeConnections.cc

</div></b>

#### Regression diffs

798 files have regression diffs
Select a representative one, for example, 5ZoneAirCooled, the diffs happen in
- audit: the number of SubTable and ColumnTags are different, due to the addition of HTML output table "AWHPs" in the equipment summary and the added entries in the "central plant" table

<img width="319" height="221" alt="image" src="https://github.com/user-attachments/assets/77a4c3a5-ba33-4db5-b4b8-af7f427af3ef" />

- tbl diff: it shows up as diff in the summary but the diff file shows 0 diff

<img width="1095" height="342" alt="image" src="https://github.com/user-attachments/assets/34f13ef3-dbd5-4d1a-95ce-1d6bad2d8f10" />

### Pull Request Author

<!-- Add to this list or remove from it as applicable.  This is a simple templated set of guidelines. -->

 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer

<!-- This will not be exhaustively relevant to every PR. -->

 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Commit:
initial commit NFP doc

Commit:
Add figures and correct equation format

yujiex commented:
> As I read the description it sounds like you intend one set of plant nodes for cooling loop connection and another set of plant nodes for heating loop connectiong. But the IDD only has one set.

I was debating on this: whether I should do separate sets of nodes for heating, cooling, and potentially domestic hot water; or resemble the actual system which just have one set of nodes (1 air 1 water). Any suggestions?

JasonGlazer commented on design/FY2025/NFP-new-air-water-heat-pump.md:
```diff
@@ -0,0 +1,1468 @@
+
+Table 2. Input fields of the heat pump objects in EnergyPlus
+
+<table class="table table-bordered table-hover table-condensed">
```
This table is almost impossible to read on GitHub since the columns are so wide and it is so tall it cannot be scrolled. I would suggest either breaking it into multiple tables or make the column width narrower by using shorter titles in each.

yujiex replied:
let me adjust the format and try to make it more readable

yujiex replied:
I reformatted the table. Hope it's more readable now.

JasonGlazer commented on design/FY2025/NFP-new-air-water-heat-pump.md:
```diff
@@ -0,0 +1,1468 @@
+## Proposed Approach ##
+
+This new feature proposes to create a new input object, HeatPump:AirToWater.
+
```
It is very difficult to compare what is being proposed to what appeared in the table of existing heat pumps because the of difficulties in that table (see previous comment). Perhaps repeat the table with the new object or else include it in the original table. Or perhaps you state that you are including all features from that table (if you are).

yujiex replied:
I will include the proposed one in the comparison table.

yujiex replied:
The proposed object is added as another row in the idd comparison table

JasonGlazer commented on design/FY2025/NFP-new-air-water-heat-pump.md:
```diff
@@ -0,0 +1,1468 @@
+- time since defrost started, timeSinceStart
+
+Note that ambient air temperature is already an existing EnergyPlus actuator. It could be used in defrost calculation as well.
+
```
Please describe the outputs that you are planning to provide. Including timestep output variables, modifications or additions to the tabular outputs, and whether any rated efficiency metrics (such as from AHRI or ISO or other standards organizations) are going to be provided.

yujiex replied:
I will add a section describing the output

yujiex replied:
I added a section listing the output variables

JasonGlazer replied:
Thanks. It would be great to add a tabular output file in the Equipment Summary report. Maybe it could include outputting a calculated rating efficiency equivalent to an AHRI or ISO test procedure (assuming there is one).

yujiex replied:
Thanks! Just added some contents to the "Outputs" section including some info on entries to be added to the Equipment Summary report.

JasonGlazer commented on design/FY2025/NFP-new-air-water-heat-pump.md:
```diff
@@ -0,0 +1,1468 @@
+boiler to produce hot water. The following is an example of the current object.
+The object specifies capacity and EIR as functions of temperature and part load
+ratio, with three performance curves.
+
```
Is this new object intending to replace the use of any existing objects and deprecate those? If not, adding a new object just seems to add some additional confusion on which one should be used. Perhaps you should include in the NFP some user oriented guidance on when each object that is mentioned in the NFP should be used including the proposed new one.

yujiex replied:
The new object can model multi-speed/variable-speed/modular products while the existing one can't. So for these cases users would want to use the new object. I will add some documentation on how to choose between them.

yujiex replied:
I added a section with notes on which model to use to model an air-to-water heat pump.

Commit:
reformat idd comparison table

Commit:
edit md: reformat idd, add sizing factor

Commit:
edit md: add output section

Commit:
edit md: notes on which model to choose

Commit:
edit md: add output table in html equipment summary

Commit:
edit md: separate output for heating cooling equip summary

Commit:
edit md: separate water inlet outlet node for different plant loop

yujiex commented:
Thanks for the feedbacks from the last technicality meetings. Regarding the few issues brought up.
- The idd object is modified to have separate water inlet outlet nodes for chilled water loop, hot water loop, and condenser loop.
- We will try to reuse existing logic/code but we're leaning towards keeping the new object with the heating and cooling component both encapsulated in this object. The new object will manage the coordination of heating and cooling operation internally.
- As for the performance curve: the 205 chiller performance map option is indeed pretty generic and clean. However manufacturers do not have this kind of data, they seem to provide data by stage (or speed level for variable-speed case). The current plant loop heat pump object performance curves cannot fully represent multi-stage products. The following is a toy example.

**Toy example**

For simplicity, we'll control theControl PLR to be the same: PLR = 0.4 (load / max capacity = 0.4)
Capacity is a function of both air and water temperature, for easy display, we'll fix the air temperature and only vary the water temperature. The figure plots the low-capacity and high-capacity heating capacity data

![capacity_curve_at_70Fair](https://github.com/user-attachments/assets/b21fafd4-9a21-4951-a62f-ef978faf7901)

As the plant loop heat pump only has one EIR curve as a function of air and water temperature, and EIR curve as a function of PLR, we want to find a PLR adjustment function f(.) such that the following holds.

$$
EIR_{high}(T_{air},T_{water}) * f(PLR) = EIR_{low}(T_{air},T_{water})
$$

As PLR is fixed to 0.4, $f(PLR)$ should be a constant, however, from the following EIR data plot. The $EIR_{high}(T_{air},T_{water}) / EIR_{low}(T_{air},T_{water})$ ratio are different for different water temperature value. We cannot find an $f(0.4)$ to make the above equation holds.

![EIR_curve_at_70Fair](https://github.com/user-attachments/assets/26209b8e-9c2a-4875-887e-f6f5ec2b0d6c)

For heating capacity, the one set of curve is also not enough: Using just one set of curve, the cycling ratio is always 0.4 (for simplicity assuming there's no minPLR limits and cycling ratio and PLR are equal) in this example for all water temperature values shown in the first figure. However, in reality, the second stage compressor will always be off and the first stage compressor will cycle. It's cycling ratio is not always constant, it varies between 0.73 and 0.75.

yujiex commented:
Re @EnergyArchmage question from last time, we have a field "Compressor Multiplier" to model modular heat pumps where several of the same AWHP objects can be used together to provide larger heating/cooling capacity. 

We will restrict the system to allow only one Air-to-Water Heat Pump (AWHP) object to serve a single hot or chilled water loop. This means that load splitting between two AWHP objects will not be permitted. The current object cannot model scenarios where two multi-stage AWHPs (AWHP-A with "Compressor Multiplier" = 2) and one variable speed compressor AWHP (AWHP-B with "Compressor Multiplier" = 1) serve the same set of heating or cooling loads, as this would require splitting loads between the two AWHPs, which is not allowed.

However, we do allow configurations where one AWHP serves the first floor and another serves the second floor, with each floor having its own dedicated plant loop.

Commit:
idd change initial commit

Commit:
idd: remove condenser inlet/outlet nodes for now

Commit:
fix idd missing "\field"

Commit:
refine idd, add input reading code and test

Commit:
add test idf to CMakeLists

Commit:
add new AWHP to BranchNodeConnections

Commit:
add doPhysics related changes, inlet outlet node input read

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
fix merge issue

Commit:
fix unit test from merging

Commit:
add new AWHP to PlantEquipTypeNames

Commit:
resolve issue from merging

Commit:
add back check for currentLoad sign, simulation unit test

Commit:
add output SetupOutputVariable

Commit:
change abs to fabs to compute absolute value for float

Commit:
add the heatpump:airtowater into the data arrays

Commit:
Add water loop type field in Plantloop

Commit:
use separate heating cooling component internally

- internally the HeatPump:AirToWater are two objects:
HeatPump:AirToWater:Heating and HeatPump:AirToWater:Cooling
- change order of items in ConnectionObjectType to match the
ConnectionObjectTypeNamesUC list

Commit:
adjust rated condition inputs to match old AWHP

Commit:
setup output variable in csv

Commit:
clang-format

Commit:
add crankcase heater calculation, remove unused include

Commit:
add entry to equipment summary central plant table

Commit:
Add AWHPs table in equipment summary html output

Commit:
add more cols to output html table AWHPs

Commit:
change capacity and power calc function

add multi-speed to power function

Commit:
revert change in Test_DoPhysics, clang format

Commit:
add rated capacity and COP for each speed level

Commit:
Add EMS variable setup, move child member out of parent class

Commit:
clang-format, setup OpMode EMS regardless, dummy EMS test OpMode EMS

Commit:
add companion coil matching

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
calc operatingMode and use EMS to control operating mode

Commit:
idd change defrost: missed max outdoor temp field, time field name

- add max outdoor temperature for defrost
- wrong field name for defrost time period fraction

Commit:
defrost related actuators and sample EMS

Commit:
clang-format

Commit:
remove overall rated ht/cl capacity, use max speed ones

Commit:
add capacity multiplier calculation

Commit:
clang-format

Commit:
initial commit of AWHP test file with only heating component

Commit:
edge case: only heating component is present

Commit:
initial commit of test file with only cooling component

Commit:
fix crankcase heater optional fields

Commit:
cooling only test idf

Commit:
Add defrost heater capacity and add more test to input reading

Commit:
unit test calcLoadSideHeatTransfer

Commit:
clang-format

Commit:
minor formatting

Commit:
unit test calcPowerUsage and fix power usage

Commit:
crankcase heater unit test

Commit:
Add input op mode ctrl option, and SingleMode calculation

Commit:
operating mode calculation with multiple unit

Commit:
add minPLR in the input

Commit:
update cyclingRatio in powerCalc function, move defrost after it

Commit:
fix heating unit test adding plantloop info in needed opmode calc

Commit:
accumulate internal arg numUnitUsed

Commit:
correct input field ordre in unit test: op mode ctrl option

Commit:
input reading unit test add multiple speed level

Commit:
test cyclingratio overwritten in calcPowerUsage

Commit:
fix issue with plant temperature run too warm

Commit:
change op mode output var to number of ht/cl unit

Commit:
heating cooling rate is the same as load side heat transfer

Commit:
apply calculated op mode, revert load profile to same as EIR_AirSource

Commit:
change defaults of min max outdoor temp to match existing object

also change internal variable to use minSourceTempLimit and
maxSourceTempLimit, same as the existing object

Commit:
fix unit test

Commit:
add input for booster mode

Commit:
add minPLR and modeCtrlOption, fix temperature too hot cold

Commit:
fix unit test, add test idf with only heating or cooling to cmakelist

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
don't multiply cycling ratio in powerUsageHigh and low

Commit:
add COP and speed level output variables

Commit:
move calcPowerUsage back to after defrost calc

as defrost multiplier is used in power calculation,
add adjustment to defrost by cycling ratio afterwords in case cycling ratio is
changed in power calculation

Commit:
interpRatio edge case, when demand > highest capacity

Commit:
differentiate speed level output: fixed vs variable speed

Commit:
use actual capacity to calculate num heating/cooling units

Commit:
add booster mode multiplier for COP

Commit:
add input-output reference

Commit:
Add COP multiplier input reading code

Commit:
allow autosize of speed 1

yzhou601 commented on design/FY2025/NFP-new-air-water-heat-pump.md:
```diff
@@ -0,0 +1,991 @@
+        \units C
+        \default 16
+        \note Enter the maximum outdoor temperature allowed for heating operation
+        \note Heating is disabled below this temperature
```
"above" this temperature?

yujiex replied:
yeah good catch. I just modified it

Commit:
clean up default fields and add unit test for defaults

Commit:
fix single-mode test cases with autosize, booster mode inputs

Commit:
remove Defrost Capacity Ratio Function of Temp Curve for now

Commit:
encoporate number of units used in EMS control of operatingMode

Commit:
remove field from test idf

Commit:
add engineering reference

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
clang-format

Commit:
remove SEER and HSPF from the equipment summary for now

Commit:
minor change of idd note

Commit:
booster mode using performance curves

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

yujiex commented:
> As I read the description it sounds like you intend one set of plant nodes for cooling loop connection and another set of plant nodes for heating loop connectiong. But the IDD only has one set.

Oh I missed this comment! The new object now has a separate pairs of nodes for chilled water loop and hot water loop

Commit:
change generator version for windows-2022 build

Commit:
revise documentation, update calc in op mode EMS and schedule

Commit:
Merge remote-tracking branch 'origin/NewAWHPobject' into NewAWHPobject

Commit:
update IO reference doc, correct output var units

Commit:
add air temp, flow rate, curve value output; update doc

Commit:
minor mod in idd: 0 to 0.0

Commit:
clang-format

yzhou601 commented on testfiles/PlantLoopHeatPump_EIR_AirSource_and_AWHP.idf:
```diff
@@ -0,0 +1,865 @@
+    , !-  Cooling Energy Input Ratio Function of Temperature Curve Name in Booster Mode
+    ; !-  Cooling Energy Input Ratio Function of PLR Curve Name in Booster Mode
+
+  HeatPump:PlantLoop:EIR:Heating,
```
Is this object still used? The name is not referenced by any other objects, but there're node connections, not sure how this works. The same question applies to `HeatPump:PlantLoop:EIR:Cooling` Can you try removing them if they are legacy objects?

yujiex replied:
I'll remove it. Thanks for catching this.

Commit:
testfile remove unused object, correct actuator comp ctrl type

Commit:
add input min max leaving water temp curve name

Commit:
add min max water temp limit curve input

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
fix doc tex math symbol

Commit:
clang-format

Commit:
input name comp multiplier to hp multiplier, minor doc updates

Commit:
minor doc update: heat pump multiplier description

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
fix in-field order, booster cooling after speed 5

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
remove unused variables

Commit:
add crankcase heater energy reporting

Commit:
tidy up output variables in test idf

Commit:
change default min air temp to -30C

Commit:
add autosizable to cooling ref capacity spd 1, change autosize calc

Commit:
increase flow rate for test idf

to reduce inlet outlet temperature difference in cooling

Commit:
fix balanced mode calculation neg unit

when cooling load is small or 0, results in negative number of cooling units

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

Commit:
shut down companion HP coil if its op mode is 0

yzhou601 commented:
Thanks for @yujiex 's great work here! I'm configuring a test case of residential AWHP with buffer tanks, and I got some confusion on the reported speed level, I discussed a bit with Yujie already, just want to also comment it here so that it's not lost track.
For test configuration, I have a simple two-stage AWHP without capacity adjustment (curve value= 1.0), with 1st speed capacity of 4000w, and 2nd speed capacity of 9000w. I'm using FixedSpeed ControlType which means it's a multi-speed system instead of a variable-speed system. And the load is a simple scheduled profile.
<img width="1526" height="791" alt="image" src="https://github.com/user-attachments/assets/bf3ce6ae-59c7-46ec-b500-7b3966317049" />
(The date is not correct on the display, the x-bar should be hours and it's a 24-hr run)
1. For the first a few hours, heat pump should be operating at the higher speed (>6000w heating loads) while the speed level is reported as 1; And for the remaining hours, it's operating at lower speed while it's reported as 0. That means, based on current implementation, 0 means the lower speed while 1 means the higher speed. It's probably better to use 0 as signal of heat pump being off, and heat pump operating levels start at 1. It's less confusing to differentiate between heat pump being off and it operating at lowest speed, and also more consistent with other objects.
2. The speeds are now reported discretely for FixedSpeed, I haven't tested VariableSpeed ControlType yet. Do you consider adding additional output variables like Cycling Ratio or Speed Ratio to report the ratio between stages, [like UnitarySystem](https://bigladdersoftware.com/epx/docs/25-1/input-output-reference/group-unitary-equipment.html#unitary-system-dx-coil-cycling-ratio)?

Again thanks for working on this and discussing details with me. Let me know if you have any questions or thoughts.

Commit:
change smallest speed level from 0 to 1

Commit:
Merge remote-tracking branch 'origin/develop' into NewAWHPobject

dareumnam commented:
> > 2. The speeds are now reported discretely for FixedSpeed, I haven't tested VariableSpeed ControlType yet. Do you consider adding additional output variables like Cycling Ratio or Speed Ratio to report the ratio between stages, [like UnitarySystem](https://bigladdersoftware.com/epx/docs/25-1/input-output-reference/group-unitary-equipment.html#unitary-system-dx-coil-cycling-ratio)?

> Re 2: There is a cycling ratio to capture the case for when the capacity at minimum speed level is larger than the demand. Would we want a speed ratio output for the demand > min-speed-capacity @dareumnam?

@yujiex I think the output (a speed ratio output for the demand > min-speed-capacity ) will definitely be helpful to the modelers, so it’s a good idea to include it as well. @mitchute do you have any other thoughts on this?

dareumnam committed:
clang format latest version

rraustad commented:
> > > 2. The speeds are now reported discretely for FixedSpeed, I haven't tested VariableSpeed ControlType yet. Do you consider adding additional output variables like Cycling Ratio or Speed Ratio to report the ratio between stages, [like UnitarySystem](https://bigladdersoftware.com/epx/docs/25-1/input-output-reference/group-unitary-equipment.html#unitary-system-dx-coil-cycling-ratio)?
> 
> > Re 2: There is a cycling ratio to capture the case for when the capacity at minimum speed level is larger than the demand. Would we want a speed ratio output for the demand > min-speed-capacity @dareumnam?
> 
> @yujiex I think the output (a speed ratio output for the demand > min-speed-capacity ) will definitely be helpful to the modelers, so it’s a good idea to include it as well. @mitchute do you have any other thoughts on this?

SpeedRatio is used for multi- or variable-speed components (e.g., the 10-speed coil model). PartLoadRatio is used for single-speed units with a cycling ratio variable to show when the unit turns on and off in a time step. You can certainly add SpeedRatio but it should be the same as PartLoadRatio so would be redundant.

Commit:
add speed ratio output

Commit:
Merge remote-tracking branch 'origin/NewAWHPobject' into NewAWHPobject

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -629,6 +628,104 @@ void EIRPlantLoopHeatPump::calcLoadSideHeatTransfer(EnergyPlusData &state, Real6
+            this->eirFuncTempCurveValue = (1 - interpRatio) * eirModifierFuncTempLow + interpRatio * eirModifierFuncTempHigh;
+            this->eirFuncPLRModifierValue = (1 - interpRatio) * eirModifierFuncPLRLow + interpRatio * eirModifierFuncPLRHigh;
+        }
+    }
```
@yujiex do you think you need to pull the powerUsage at lines 702 and 703 into this if/else block? why set speed level and performance curves at the high speed for CompressorControlType::FixedSpeed and interpolated otherwise while the power calculation always interpolates?

yujiex replied:
for the multi-stage fixed-speed case, when the demand is between speed level n and speed level n-1, the heat pump runs p% time at speed n and (1-p)% time at speed n - 1. 
for the variable speed case, it interpolates the performance curve to mimic modulating to a intermediate speed to satisfy the demand

I felt the two should theoretically result in the same power consumption? Like if the two speed level has capacity 100W and 200W and that the demand is 150W for example, then in the first case, it will have 50% time running at the lower speed and 50% time running at the higher speed; in the second case, it will interpolate the two speed levels with interpolation factor of 0.5. In both cases, they'll have the same power consumption: average of the power at the two speed level?

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -629,6 +628,104 @@ void EIRPlantLoopHeatPump::calcLoadSideHeatTransfer(EnergyPlusData &state, Real6
     // currentLoad will be met and there should? be some adjustment based on outlet water temp limit?
 }
 
+void HeatPumpAirToWater::calcLoadSideHeatTransfer(EnergyPlusData &state, Real64 const availableCapacity, Real64 const currentLoad)
```
This may be out or scope for this new model but all plant equipment models should be passing back the load met (Real64 &currentLoad) to the plant manager (i.e., returned from the simulate function call to the plant manager) so the plant knows what the remaining load is. If the load is greater than the equipment capacity then the plant will have a remaining load to meet. All the simulate functions should look like this: `simulate(state, , , Real64 &curLoad) {}`

yujiex replied:
I see. Although in `void EIRPlantLoopHeatPump::simulate()`, it seems like `CurLoad` is not updated in the function: in in both `setOperatingFlowRatesASHP(..., CurLoad)`, and `doPhysics(state, CurLoad)`, `CurLoad` is passed by value.

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -629,6 +628,104 @@ void EIRPlantLoopHeatPump::calcLoadSideHeatTransfer(EnergyPlusData &state, Real6
+    // calculate power usage from EIR curves
+    Real64 eirModifierFuncTempLow = 1.0;
+    Real64 eirModifierFuncPLRLow = 1.0;
+    if (speedLevel > 0) {
```
I noticed at the bottom of this function is `this->speedLevel += 1;` and that this conditional and the curve indexes below use a local variable `speedLevel`. This local variable is always = the operating speed. So what does the `this->speedLevel += 1;` at the bottom of the function do?

yujiex replied:
`this->speedLevel += 1` is for reporting. In the internal calculation, the speed level is 0-indexed. But to follow the reporting convention, speed level need to start from 1. So there is the shift in `this->speedLevel`.

Commit:
fix test idf output var ;

rraustad reviewed (approved):
All questions were sufficiently answered. @mitchute this looks ready.

mitchute committed:
update idf version numbers

mitchute commented:
I bumped the version numbers up to current. We can merge this after CI wraps up one last time.

dareumnam reviewed (approved):
@rraustad, thanks so much for your review and input.
I built it locally, ran the newly added unit tests (all passed), and ran the created test files successfully, confirming the results match what Yujie explained. CI checks look good, diffs are explained and verified, comments have been addressed, and all related documentation has been updated. This one looks all set to go. @mitchute

mitchute committed:
fix output var spelling, format input files

mitchute commented:
OK, I made another quick review and made one final commit. We'll let CI finish, then merge this.

mitchute commented:
Been waiting most of the day, but I think Decent CI is hung up here. Merging and moving on. Thanks all.

jmarrec commented on testfiles/PlantLoopHeatPump_EIR_AirSource_and_AWHP.idf:
```diff
@@ -0,0 +1,711 @@
+    0.0,                     !- Coefficient2 x
+    0.0,                     !- Coefficient3 x**2
+    0.0,                     !- Coefficient4 y
+    0.0,                     !- Coefficient5 y**2
+    0.0,                     !- Coefficient6 x*y
+    5.0,                     !- Minimum Value of x
+    10.0,                    !- Maximum Value of x
+    24.0,                    !- Minimum Value of y
+    35.0,                    !- Maximum Value of y
+    ,                        !- Minimum Curve Output
+    ,                        !- Maximum Curve Output
+    Temperature,             !- Input Unit Type for X
+    Temperature,             !- Input Unit Type for Y
+    Dimensionless;           !- Output Unit Type
+
+  Curve:Biquadratic,
+    EIRCurveFuncTemp,        !- Name
+    1.0,                     !- Coefficient1 Constant
+    0.0,                     !- Coefficient2 x
+    0.0,                     !- Coefficient3 x**2
+    0.0,                     !- Coefficient4 y
+    0.0,                     !- Coefficient5 y**2
+    0.0,                     !- Coefficient6 x*y
+    5.0,                     !- Minimum Value of x
+    10.0,                    !- Maximum Value of x
+    24.0,                    !- Minimum Value of y
+    35.0,                    !- Maximum Value of y
+    ,                        !- Minimum Curve Output
+    ,                        !- Maximum Curve Output
+    Temperature,             !- Input Unit Type for X
+    Temperature,             !- Input Unit Type for Y
+    Dimensionless;           !- Output Unit Type
+
+  Curve:Quadratic,
+    EIRCurveFuncPLR,         !- Name
+    1.0,                     !- Coefficient1 Constant
+    0.0,                     !- Coefficient2 x
+    0.0,                     !- Coefficient3 x**2
+    0.0,                     !- Minimum Value of x
+    1.0;                     !- Maximum Value of x
```
Sorry for jumping late, i just noticed this PR dropping in develop yesterday and I am interested in the feature, as I have had to model AWHP before including some systems like the one mentioned in the PR's original post.

I'm a bit confused by these curves, and they seem really generic to mimic a real world system.

I'm sure the testfiles added are nice to ensure that a simulation runs and for coverage, but it's also an Example File, so does it mimic real world operation?

jmarrec commented on testfiles/PlantLoopHeatPump_EIR_AirSource_and_AWHP_cooling_only.idf:
```diff
@@ -0,0 +1,497 @@
+    Verbose,                 !- Internal Variable Availability Dictionary Reporting
+    Verbose;                 !- EMS Runtime Language Debug Output Level
+
+  EnergyManagementSystem:Actuator,
+    opModeAct_cooling,       !- Name
+    test_AWHP,               !- Actuated Component Unique Name
+    HeatPump:AirToWater:Cooling,  !- Actuated Component Type
+    Operating Mode;          !- Actuated Component Control Type
+
+  EnergyManagementSystem:OutputVariable,
+    Operating Model Output Cooling,  !- Name
+    opModeAct_cooling,       !- EMS Variable Name
+    Averaged,                !- Type of Data in Variable
+    SystemTimeStep,          !- Update Frequency
+    ,                        !- EMS Program or Subroutine Name
+    ;                        !- Units
+
+  EnergyManagementSystem:ProgramCallingManager,
+    Change_OpMode_EMS_Program_Manager,  !- Name
+    InsideHVACSystemIterationLoop,  !- EnergyPlus Model Calling Point
+    Change_OpMode_EMS_Program;  !- Program Name 1
+
+  EnergyManagementSystem:Program,
+    Change_OpMode_EMS_Program,  !- Name
+    IF (Hour <= 12),         !- Program Line 2
+    SET opModeAct_cooling = 0,  !- <none>
+    ELSE,                    !- <none>
+    SET opModeAct_cooling = 1,  !- <none>
+    ENDIF;                   !- <none>
```
I don't get this EMS program here?

Again, this ends up being an Example File.

yujiex replied:
This is some example EMS code for controlling the operating mode. Do you feel it's too simplified and better be removed? I can update the example in a followup PR.

jmarrec commented on testfiles/PlantLoopHeatPump_EIR_AirSource_and_AWHP_cooling_only.idf:
```diff
@@ -0,0 +1,497 @@
+! PlantLoopHeatPump_EIR_AirSource.idf
+!
+! Basic file description:
+! The EIR formulated air-to-water heat pump is demonstrated using a simple collection of plant loops.  Building loads
+! are represented using Plant Load Profile objects
+! Run:      1 design day.
+! Building: None.
+! System:   None.
+! Plant:    PLANT LOAD PROFILE with District Heating and Cooling as necessary.
```
The three testfiles added have exactly the same description. I feel like there should be be some difference here.

And the mention of the special shenanigans like the EMS program would probably be good for the heating/cooling only ones.

yujiex replied:
Indeed. I forgot to update the description. I will modify them.

dareumnam replied:
Thanks for catching this and for all your review! @jmarrec

jmarrec reviewed (commented):
Just have a couple of questions / suggestions. Sorry for dropping in late.

I'm guessing you tested this with real world manufacturer data?

A datasets/ file would be nice, or at least one example file should use proper data.

yujiex commented:
> Just have a couple of questions / suggestions. Sorry for dropping in late.
> 
> I'm guessing you tested this with real world manufacturer data?
> 
> A datasets/ file would be nice, or at least one example file should use proper data.

Yes, a Daikin engineer is using their performance data to test the model. But they don't want to reveal their product data, so the one here uses very generic setting in the curves. I'll try to create another example with more realistic data. I'll create a followup PR to add it in.
